### PR TITLE
SALTO-1102-Support iterative validation errors calc

### PR DESCRIPTION
**Pull request by noamhammer**
Changed the way we calculate validation errors - on every change we calculate the validation errors on:
- elements in the changed files
- elements that already got validation error
In addition, we add unresolved references errors on elements with reference to deleted elements
Moreover, we calculate the validation errors every time we open a new file, on the new file.

---
Release Notes: Improve the performance of the validation errors calculation and therefore, the performance of the extension.

### CODE REVIEW

File: packages/lang-server/src/workspace.ts [hunk 1]
```diff
@@ -96,20 +100,117 @@ export class EditorWorkspace {
...
+      .map(e => e.elemID)
+  }
+
+  private async getUnresolvedRefOfFile(target: ElemID, filename: string):
+  Promise<errors.UnresolvedReferenceValidationError[]> {
+    const elements = (await this.workspace.getParsedNaclFile(
+      this.workspaceFilename(filename)
+    ))?.elements ?? []
+    const ids = new Set<string>()
+    const transformFunc = ({ value, path: elemPath }: TransformFuncArgs): Value => {
```
Comment (netama) [inline packages/lang-server/src/workspace.ts:117]: nit: maybe give the function a less generic name? :)

File: packages/lang-server/src/workspace.ts [hunk 2]
```diff
@@ -96,20 +100,117 @@ export class EditorWorkspace {
...
+      if (isReferenceExpression(value) && elemPath) {
+        if (target.isEqual(value.elemId) || target.isParentOf(value.elemId)) {
+          ids.add(elemPath.getFullName())
+        }
+      }
+      return value
+    }
+    elements.forEach(element => {
+      if (!isContainerType(element)) {
```
Comment (netama) [inline packages/lang-server/src/workspace.ts:126]: can you please explain why this doesn't need to go over containers (or their inner types)? is it because the assumption is that the inner type is the responsibility of the file where it's defined?

File: packages/lang-server/src/workspace.ts [hunk 3]
```diff
@@ -96,20 +100,117 @@ export class EditorWorkspace {
...
+      }
+    })
+    return [...ids].map(id => new errors.UnresolvedReferenceValidationError(
+      { elemID: ElemID.fromFullName(id), target }
+    ))
+  }
+
+  private async getUnresolvedRefForElement(element: Element):
+  Promise<errors.UnresolvedReferenceValidationError[]> {
+    const referencedFiles = await this.getElementReferencesToFiles(element.elemID)
```
Comment (netama) [inline packages/lang-server/src/workspace.ts:137]: this has `reference*d*Files` and `getElementReference*sTo*Files`, any chance one of them should be changed? or does the distinction not matter in this context?

File: packages/lang-server/src/workspace.ts [hunk 4]
```diff
@@ -96,20 +100,117 @@ export class EditorWorkspace {
...
+    return (await Promise.all(
+      referencedFiles.map(file => this.getUnresolvedRefOfFile(element.elemID, file))
+    )).flat()
+  }
+
+  private async getValidationErrors(files: string[], changes: Change<Element>[]):
+  Promise<errors.ValidationError[]> {
+    // We update the validation errors iterativly by validate the following elements:
```
Comment (netama) [inline packages/lang-server/src/workspace.ts:145]: nit: `validate` -> `validating`

File: packages/lang-server/src/workspace.ts [hunk 5]
```diff
@@ -96,20 +100,117 @@ export class EditorWorkspace {
...
+    //   - all the elements in the changed files
+    //   - all the elements to includes reference expressions to them
```
Comment (netama) [inline packages/lang-server/src/workspace.ts:147]: I don't understand this sentence :) should this have been something like `all the elements that the changed files include reference expressions to them` / `all the elements that are referenced by elements in the changed files`?

File: packages/lang-server/src/workspace.ts [hunk 6]
```diff
@@ -96,20 +100,117 @@ export class EditorWorkspace {
...
+      (await this.errors()).validation
+        .forEach(ve => {
+          elementNamesToValidate.add(ve.elemID.createTopLevelParentID().parent.getFullName())
+        })
+    }
+    const parsedNaclFiles = (await Promise.all(files.map(
+      file => this.workspace.getParsedNaclFile(this.workspaceFilename(file))
+    ))).filter(values.isDefined)
+    parsedNaclFiles.forEach(parsed => {
+      (parsed?.elements ?? [])
```
Comment (netama) [inline packages/lang-server/src/workspace.ts:160]: nit: do we still need `?.` and the fallback after filtering `values.isDefined`?

File: packages/lang-server/src/workspace.ts [hunk 7]
```diff
@@ -96,20 +100,117 @@ export class EditorWorkspace {
...
+    const elementsToValidate = (await Promise.all([...elementNamesToValidate]
+      .map(name => this.workspace.getElement(ElemID.fromFullName(name)))))
+      .flat()
+      .filter(values.isDefined)
+    const validationErrors = validateElements(
+      elementsToValidate, await this.workspace.elements()
+    )
+    const unresolvedReferences = (await Promise.all(
+      changes
+        .filter(isRemovalChange)
```
Comment (netama) [inline packages/lang-server/src/workspace.ts:172]: not 100% sure, but I think we may need to look at modification changes (and sub-element parts) as well, because we now have reference expressions pointing into values inside instances which I _think_ will not get their own removal change? (@nitzan-haim added that capability recently - you can see an example in https://github.com/salto-io/salto/blob/f6a77b8e7dce2fee4046b5e7da082be62a5a9ce8/packages/salesforce-adapter/test/transformers/transformer.test.ts#L1508-L1515 )

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 8]
```diff
@@ -44,6 +44,7 @@ const { withLimitedConcurrency } = promises.array
 const log = logger(module)
 
 
+type References = { referenced: ElemID[]; referencesTo: ElemID[] }
```
Comment (netama) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:47]: naming nit: `referencesTo` is not that clear to me (and because of that I'm not sure this suggestion is correct :) ) - would it make sense to rename it to something like `referencedAncestors` / `ancestorsOfReferenced`? does it contain all the parents of the referenced elements?

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 9]
```diff
@@ -124,7 +129,7 @@ const getElementReferenced = (element: Element): ElemID[] => {
         value.elemId.idType,
         ...valueIDPath.slice(0, index + 1)
       ))
-      referenced.push(parent, ...nestedIds)
+      referencesTo.push(parent, ...nestedIds)
```
Comment (netama) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:132]: shouldn't `value.elemId` itself still be added to `referenced` (I think until now it was just included as one of the `nestedIds`)?

**Review by netama — COMMENTED** [review]
I don't have all the context about this change so probably best if Ori/Roi review - but added some questions (one I think may be more important than the others - around handling sub-element references)

File: packages/lang-server/src/workspace.ts (hunk 7 above)
Comment (netama) [reply packages/lang-server/src/workspace.ts:172]: never mind, I think removals should be enough...

File: packages/lang-server/src/workspace.ts (hunk 5 above)
Comment (noamhammer) [reply packages/lang-server/src/workspace.ts:147]: rewrote the comment

File: packages/lang-server/src/workspace.ts (hunk 2 above)
Comment (noamhammer) [reply packages/lang-server/src/workspace.ts:126]: Yes, IIUC, it's the responsibility of the inner type. In addition, that's also the current behavior(As you can see in `getElemIDUsages`)

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts (hunk 8 above)
Comment (noamhammer) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:47]: `referencesTo` means elements with references expressions to the element/s. Any suggestions for a better name?

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts (hunk 9 above)
Comment (noamhammer) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:132]: `referencesTo` are elements that include reference expression to this element and referenced is all the referenced to the element without `referencesTo`. Does it makes sense to you?

File: packages/lang-server/src/workspace.ts [hunk 10]
```diff
@@ -96,20 +100,116 @@ export class EditorWorkspace {
...
-      // Now add the waiting changes
-      if (!_.isEmpty(opNaclFiles) && this.workspace) {
-        await this.workspace.setNaclFiles(..._.values(opNaclFiles))
+      if (_.isUndefined(this.wsErrors)) {
+        await this.update([...opDeletes], opNaclFiles)
+        await this.errors()
+      } else {
+        const changes = await this.update([...opDeletes], opNaclFiles)
+        const validation = await this.getValidationErrors(
+          [...opDeletes, ...Object.keys(opNaclFiles)], changes
```
Comment (ori-moisis) [inline packages/lang-server/src/workspace.ts:205]: if we call this with the deleted files after the workspace is updated, won't we miss the IDs of the elements in files that were deleted?
say I have element A split between two files:
```
// File 1
bla A {
  val = 1
}
// File 2
bla A {
  foo = 2
}
```
lets say `val` is a required field and I delete File 1.
In that cause I won't know that I should validate `A` because we remove the file before we check which elements are in it.
Am I missing something?

File: packages/lang-server/src/workspace.ts [hunk 11]
```diff
@@ -183,11 +283,41 @@ export class EditorWorkspace {
...
     return (await this.workspace.getElementNaclFiles(id))
       .map(filename => this.editorFilename(filename))
   }
 
+  async validateFiles(filenames: string[]): Promise<errors.Errors> {
+    if (_.isUndefined(this.wsErrors)) {
+      return this.errors()
+    }
+    const elements = (await this.elementsInFiles(filenames)).map(e => e.getFullName())
+    const currentErrors = await this.errors()
```
Comment (ori-moisis) [inline packages/lang-server/src/workspace.ts:301]: this is really confusing that we use `this.errors()` in both cases (when `this.wsErrors` is undefined and when it isn't).
I am also not sure it is OK to just call this function that relies on the workspace to remain the same across many async calls without "locking" like we do for `runAggregatedSetOperation `.
the workspace might change between the check that `this.wsErrors !== undefined` and the call to `this.errors()` and it can change again between every call to `getElement` a few lines down from here

File: packages/lang-server/src/workspace.ts [hunk 12]
```diff
@@ -183,11 +283,41 @@ export class EditorWorkspace {
...
+    const validation = currentErrors.validation
+      .filter(e => !elements.includes(e.elemID.createTopLevelParentID().parent.getFullName()))
+      .concat(validateElements(
+        (await Promise.all(elements.map(e => this.workspace.getElement(ElemID.fromFullName(e)))))
+          .filter(values.isDefined),
+        await this.workspace.elements(),
```
Comment (ori-moisis) [inline packages/lang-server/src/workspace.ts:307]: copy-pasted logic from `getValidationErrors`?
why re-write and not re-use?

File: packages/lang-server/src/workspace.ts [hunk 13]
```diff
@@ -96,20 +100,116 @@ export class EditorWorkspace {
...
+    ))).filter(values.isDefined)
+    parsedNaclFiles.forEach(parsed => {
+      parsed.elements.forEach(elem => elementNamesToValidate.add(elem.elemID.getFullName()))
+    })
+    const elementsToValidate = (await Promise.all([...elementNamesToValidate]
+      .map(name => this.workspace.getElement(ElemID.fromFullName(name)))))
+      .flat()
+      .filter(values.isDefined)
+    const validationErrors = validateElements(
+      elementsToValidate, await this.workspace.elements()
```
[the remark this answers is not in the document]
Comment (noamhammer) [reply packages/lang-server/src/workspace.ts:167]: let N be the number of the workspace' elements and let M be the number of the elements to validate
Your solution is O(N*M) and mine is O(M).
Am I wrong?

[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/lang-server/src/workspace.ts:181]: assuming `elementNamesToValidate` is a hash set (so `.has` is O(1)) my suggestion is O(n).
and since the current code gets all the workspace elements with `await this.workspace.elements()` it is O(n) as well.

It is true that the current solution makes one pass less over the elements so my suggestion will be a little slower and I took that into account when I suggested it. however, I don't think it will be noticeably slower and it makes the code so much simpler that I think it is worth it - especially the part where we can remove the changes in the workspace implementation / interface which can save us some really annoying conflicts with the lazy elements branch

File: packages/lang-server/test/usage.test.ts [hunk 14]
```diff
@@ -83,18 +84,21 @@ describe('Test go to definitions', () => {
...
+    const context = await getPositionContext(workspace, naclFileName, pos)
     const defs = await provideWorkspaceReferences(workspace, token, context)
     expect(getRefLines(defs)).toEqual([208])
   })
 
-  it('should find (goto) refrences of a nested (salto) references', async () => {
+  // TODO: There is a bug that prevents us from add a reference from instance to it's type.
+  //  Once we fix this bug, we should unskip this test.
+  // eslint-disable-next-line jest/no-disabled-tests
+  it.skip('should find (goto) refrences of a nested (salto) references', async () => {
```
Comment (ori-moisis) [inline packages/lang-server/test/usage.test.ts:95]: ummm, so, this test did pass before, right? so if there is a bug now, and the test passes currently, that means the bug was added in this PR, no?

File: packages/workspace/src/workspace/workspace.ts [hunk 15]
```diff
@@ -70,6 +70,11 @@ export type WorkspaceComponents = {
   serviceConfig: boolean
 }
 
+type WorkspaceState = {
+  mergeErrors: MergeError[]
+  elements: Record<string, Element>
+}
```
Comment (ori-moisis) [inline packages/workspace/src/workspace/workspace.ts:76]: why is this change needed?
seems like most of the usages changed to be more complicated then they used to be...

File: packages/lang-server/src/workspace.ts [hunk 16]
```diff
@@ -96,20 +100,130 @@ export class EditorWorkspace {
...
+        changes.map(c => getChangeElement(c).elemID)
+      ).map(elemID => elemID.getFullName())
+    )
+    const validationErrors = await this.validateElements(elementNamesToValidate)
+    const removalChangesOfTopLevels = changes
+      .filter(isRemovalChange)
+      .map(c => getChangeElement(c).elemID)
+    const removalChangesOfNonTopLevels = changes
+      .filter(isModificationChange)
+      .filter(isTypeOrInstanceChange)
```
[the remark this answers is not in the document]
Comment (noamhammer) [reply packages/lang-server/src/workspace.ts:186]: Field

File: packages/lang-server/test/usage.test.ts (hunk 14 above)
Comment (noamhammer) [reply packages/lang-server/test/usage.test.ts:95]: https://salto-io.atlassian.net/browse/SALTO-1130

File: packages/lang-server/src/workspace.ts [hunk 17]
```diff
@@ -96,20 +100,130 @@ export class EditorWorkspace {
...
+  private async getValidationErrors(files: string[], changes: Change<Element>[]):
+  Promise<errors.ValidationError[]> {
+    // We update the validation errors iterativly by validating the following elements:
+    //   - all the elements in the changed files
+    //   - all the elements that includes references expression to removed elements
+    //   - all the elements that currently got validation errors
+    const elementsWithValidationErrors = this.wsErrors === undefined
+      ? []
+      : (await this.wsErrors).validation.map(error => error.elemID.createTopLevelParentID().parent)
+    const elementsInChangedFiles = await this.elementsInFiles(files)
```
[the remark this answers is not in the document]
Comment (noamhammer) [reply packages/lang-server/src/workspace.ts:172]: I thought it's ok since I added `changes.map(c => getChangeElement(c).elemID)` to the elements to validate - so each removed element will also be validated. Doesn't it ok? If it's still not ok let's talk f2f

File: packages/lang-server/src/workspace.ts [hunk 18]
```diff
@@ -96,20 +100,130 @@ export class EditorWorkspace {
...
 
+  private async elementsInFiles(filenames: string[]): Promise<ElemID[]> {
+    return (await Promise.all(
+      filenames.map(f => this.workspace.getParsedNaclFile(this.workspaceFilename(f)))
+    )).filter(values.isDefined)
+      .flatMap(parsed => parsed.elements)
+      .map(e => e.elemID)
+  }
+
+  private async getUnresolvedRefOfFile(filename: string, references: ElemID[]):
```
[the remark this answers is not in the document]
Comment (noamhammer) [reply packages/lang-server/src/workspace.ts:111]: You are correct. It's bad naming.

File: packages/lang-server/src/workspace.ts [hunk 19]
```diff
@@ -96,20 +100,130 @@ export class EditorWorkspace {
...
+    const elements = (await this.workspace.getParsedNaclFile(
+      this.workspaceFilename(filename)
+    ))?.elements ?? []
+    const validationErrors: errors.UnresolvedReferenceValidationError[] = []
+    const getReferenceExpressions = ({ value, path: elemPath }: TransformFuncArgs): Value => {
+      if (isReferenceExpression(value) && elemPath) {
+        references.forEach(ref => {
+          if (ref.isEqual(value.elemId) || ref.isParentOf(value.elemId)) {
+            validationErrors.push(new errors.UnresolvedReferenceValidationError(
+              { elemID: elemPath, target: ref }
```
[the remark this answers is not in the document]
Comment (noamhammer) [reply packages/lang-server/src/workspace.ts:122]: As always, you are correct

File: packages/workspace/src/workspace/workspace.ts (hunk 15 above)
Comment (noamhammer) [reply packages/workspace/src/workspace/workspace.ts:76]: There are places that need to get specific element from elements (like `getValue`) and it's more efficient to do it like that. Do you think it is not worth it?

File: packages/lang-server/src/workspace.ts [hunk 20]
```diff
@@ -96,20 +100,130 @@ export class EditorWorkspace {
...
+    ).flatten()
+      .groupBy('file')
+      .mapValues(val => _.uniqBy(val.map(ref => ref.id), e => e.getFullName()))
+      .value()
+    return (await Promise.all(
+      Object.entries(fileToReferencedIds)
+        .map(([filename, elemIds]) =>
+          this.getUnresolvedRefOfFile(filename, elemIds))
+    )).flat()
+  }
```
[the remark this answers is not in the document]
Comment (noamhammer) [reply packages/lang-server/src/workspace.ts:154]: - re: complexity - I did it for optimization purposes (you can in `getUnresolvedRefOfFile` function that we only compare the relevant elements for the file in the transform element). If you think that this optimization is irrelevant and we better remove it, please tell me and I will do so.
- re: `createTopLevelParentID` - You are correct. I removed the `createTopLevelParentID` part and just passed `id` and I believe that it revealed a bug (IMO, we build the `referenced` list in `ParsedNaclFile` wrongly) - added a fix for it for now.

File: packages/lang-server/src/workspace.ts [hunk 21]
```diff
@@ -188,6 +302,31 @@ export class EditorWorkspace {
...
   }
 
+  private async validateFilesImpl(filenames: string[]): Promise<errors.Errors> {
+    if (_.isUndefined(this.wsErrors)) {
+      return this.errors()
+    }
+    const currentErrors = await this.wsErrors
+    const elements = new Set((await this.elementsInFiles(filenames)).map(e => e.getFullName()))
+    const validation = currentErrors.validation
+      .filter(e => !elements.has(e.elemID.createTopLevelParentID().parent.getFullName()))
```
[the remark this answers is not in the document]
Comment (noamhammer) [reply packages/lang-server/src/workspace.ts:312]: That's the flow for switching files, the other flow is for editing. I am not sure that we want to run validation on all the current errors when we switching files. Do you think we should?

File: packages/lang-server/src/workspace.ts [hunk 22]
```diff
@@ -96,20 +100,130 @@ export class EditorWorkspace {
...
+    const unresolvedReferences = await this.getUnresolvedRefForElement(
+      [...removalChangesOfTopLevels, ...removalChangesOfNonTopLevels]
+    )
+    const unresolvedReferencesElemIDs = new Set(
+      unresolvedReferences.map(e => e.elemID.getFullName())
+    )
+    return validationErrors
+      .filter(e =>
+        !validator.isUnresolvedRefError(e)
+        || !unresolvedReferencesElemIDs.has(e.elemID.getFullName()))
```
[the remark this answers is not in the document]
Comment (noamhammer) [reply packages/lang-server/src/workspace.ts:199]: I am not sure that it would make the code clearer. The elements that we need to filter out are of the elements in the referenced files to the removed elements. Which means that in order to do it, we need to pass it to the inner functions, and from first look (but I might be wrong) I am not sure it's that straight forward. But again, if you still think it should be done, please tell and I will do so

File: packages/lang-server/src/workspace.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/lang-server/src/workspace.ts]: I think the additional work required to calculate the significantly more complex and bigger data structure is much more than what you potentially save from checking a slightly larger set.
unless I am completely missing something here and this makes a noticeable difference in performance I think we shouldn't over-complicate our code

File: packages/lang-server/src/workspace.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/lang-server/src/workspace.ts:186]: how would it be a field?
and if it is a field, why is it ok to ignore?

File: packages/lang-server/src/workspace.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/lang-server/src/workspace.ts:172]: > If we don't need to consider the elements from before the update, are we sure we need to consider the elements from after the update?

doesn't make sense to me that we need one and not the other, so if you are confident we don't need to check the elements from before the update, are you sure we need to check the elements from after the update?

File: packages/lang-server/src/workspace.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/lang-server/src/workspace.ts:312]: the question was why is there a difference.
if you think we don't need to run validation in this flow, then why do we need to run it in the other flow? what is the difference that makes it ok in one flow but not the other?

File: packages/workspace/src/workspace/workspace.ts (hunk 15 above)
Comment (ori-moisis) [reply packages/workspace/src/workspace/workspace.ts:76]: so it really isn't related to the purpose of this PR? that is what confused me, I couldn't understand why this change had to happen in this PR.
I don't mind it, just couldn't understand why it was needed so I assumed I missed something important

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 23]
```diff
@@ -112,36 +111,47 @@ const getTypeOrContainerTypeID = (typeElem: TypeElement): ElemID => (isContainer
...
   }
-  return _.uniq(referenced)
+  return referenced
+}
+
+const getElementsReferences = (elements: Element[]): ElemID[] => {
+  const referenced = new Set<string>()
+  elements.forEach(elem => {
+    getElementReferenced(elem).forEach(r => referenced.add(r))
+  })
```
Comment (ori-moisis) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:153]: style nit - avoid side effects when not needed
```suggestion
  const referenced = new Set(wu(elements).map(getElementReferenced))
```

File: packages/lang-server/test/usage.test.ts (hunk 14 above)
Comment (ori-moisis) [reply packages/lang-server/test/usage.test.ts:95]: k

**Review by ori-moisis — APPROVED** [review]
I still don't really understand everything here but it seems like you are pretty confident that it works so it would probably be easier to merge this and fix anything we missed if/when we encounter it later

**Comment by ori-moisis** [comment]
btw, can we do something about the coverage? a PR that adds code usually shouldn't decrease coverage

File: packages/lang-server/test/usage.test.ts (hunk 14 above)
Comment (noamhammer) [reply packages/lang-server/test/usage.test.ts:95]: We should either change the test or fix the code - depends on the desired behavior

**Comment by noamhammer** [comment]
@ori-moisis It seems like you didn't really want to approve this PR if this is your comment, so don't worry, I am not going to merge it. If there is something that you still don't understand, tell me and I will do my best to answer/fix it.

> I still don't really understand everything here but it seems like you are pretty confident that it works so it would probably be easier to merge this and fix anything we missed if/when we encounter it later

**Comment by ori-moisis** [comment]
these are the the things I still don't understand:
https://github.com/salto-io/salto/pull/1699#discussion_r559456454
https://github.com/salto-io/salto/pull/1699#discussion_r559450682
https://github.com/salto-io/salto/pull/1699#discussion_r559448897

but I did mean it when I said that if you are confident this works then we can merge it.
the effect of a bug in any of the things I don't understand should be contained to the language server so it shouldn't be too bad even if we did miss something.
